### PR TITLE
Allow setting mutex locking/unlocking callbacks before trace init.

### PR DESCRIPTION
**Pull request by tommikas**
Calling mbed_trace_init() set the callbacks to NULL so they had to be set after initializing the library.
This left a small window of time during which traces could happen without any thread safety being in place.

### CODE REVIEW

**Comment by tommikas** [comment]
@RomanSaveljev @kjbracey-arm Could you review please.

File: source/mbed_trace.c [hunk 1]
```diff
@@ -162,9 +162,6 @@ int mbed_trace_init(void)
     m_trace.suffix_f = 0;
     m_trace.printf = mbed_trace_default_print;
     m_trace.cmd_printf = 0;
-    m_trace.mutex_wait_f = 0;
-    m_trace.mutex_release_f = 0;
-    m_trace.mutex_lock_count = 0;
```
Comment (RomanSaveljev) [inline source/mbed_trace.c]: Ok, does not belong to this PR, but I would use here something like "NULL-object" pattern to avoid `if (m_trace.mutex_wait_f) {...}` everywhere

**Comment by RomanSaveljev** [comment]
Do we need a mutex for `mbed_trace_init` and then a mutex for assigning a mutex function? :)

File: source/mbed_trace.c (hunk 1 above)
Comment (tommikas) [inline source/mbed_trace.c]: Yep, that would be a good idea.

**Comment by tommikas** [comment]
Currently the mutexes are only meant to protect the actual trace calls. Initialization and configuration aren't protected.

**Comment by RomanSaveljev** [comment]
Needs to be documented carefully, because of the "creativity window" we leave open for the user.

LGTM

**Comment by tommikas** [comment]
I'll add a little more documentation.

**Comment by tommikas** [comment]
@SeppoTakalo Could you take a look at this please.

**Comment by SeppoTakalo** [comment]
+1
